src/reconcile/rec_visibility.c

```diff
@@ -252,6 +257,81 @@ __timestamp_out_of_order_fix(WT_SESSION_IMPL *session, WT_TIME_WINDOW *select_tw
     return (false);
 }
 
+/*
+ * __rec_validate_upd_chain --
+ *     Check the update chain for conditions that would prevent it's insertion into the history
+ *     store. Return EBUSY if the update chain cannot be inserted into the history store at this
+ *     time.
+ */
+static int
+__rec_validate_upd_chain(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_UPDATE *select_upd,
+  WT_TIME_WINDOW *select_tw, WT_CELL_UNPACK_KV *vpack)
+{
+    WT_DECL_RET;
+    WT_UPDATE *upd;
+    wt_timestamp_t current_ts;
+    bool seen_ondisk_value;
+
+    upd = NULL;
+    current_ts = WT_TS_NONE;
+    seen_ondisk_value = false;
+
+    /*
+     * If we're not in eviction any history store insertion is fine, if we are in eviction and a
+     * checkpoint isn't running then history store insertions are also fine.
+     */
+    if (!F_ISSET(r, WT_REC_EVICT) || !F_ISSET(r, WT_REC_CHECKPOINT_RUNNING))
+        return (0);
+
+    /*
+     * The selected time window may contain information that isn't visible given the selected
+     * updated, as such we have to check it separately. This is true when there is a tombstone ahead
+     * of the selected update.
+     */
+    if (select_tw->stop_ts < select_tw->start_ts)
+        WT_ERR(EBUSY);
+
+    /*
+     * There is no selected update to go to disk as such we don't need to check the updates
+     * following it.
+     */
+    if (select_upd == NULL)
+        return (0);
+
+    /* Set our variables correctly now that we are sure there's an update. */
+    upd = select_upd->next;
+    current_ts = select_upd->start_ts;
+
+    /* Loop forward from update after the selected on-page update. */
+    for (; upd != NULL; upd = upd->next) {
+        /* Validate that the updates older than us have older timestamps. */
+        if (current_ts < upd->start_ts)
+            WT_ERR(EBUSY);
+
+        /*
+         * If we see transactions id's that belong to the on disk value we don't need to check it
+         * later.
+         */
+        if (vpack != NULL &&
+          (upd->txnid == vpack->tw.stop_txn || upd->txnid == vpack->tw.start_txn))
+            seen_ondisk_value = true;
+        current_ts = upd->start_ts;
+    }
+
+    /* Check that the on-page time window isn't out-of-order. */
+    if (!seen_ondisk_value && vpack != NULL &&
+      (current_ts < vpack->tw.start_ts ||
+        (vpack->tw.stop_ts != WT_TS_NONE && vpack->tw.stop_ts != WT_TS_MAX &&
```

Comment:
Why this check is here `vpack->tw.stop_ts != WT_TS_NONE`?

Reply:
`vpack->tw.stop_ts != WT_TS_MAX` can be replaced by `WT_TW_HAS_STOP(session, &tw)`